cmd/juju/environment/create.go

```diff
@@ -0,0 +1,210 @@
+
+package environment
+
+import (
+	"strings"
+
+	"github.com/juju/cmd"
+	"github.com/juju/errors"
+	"github.com/juju/utils/keyvalues"
+	"gopkg.in/yaml.v1"
+	"launchpad.net/gnuflag"
+
+	"github.com/juju/juju/api/environmentmanager"
+	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/juju/cmd/envcmd"
+	"github.com/juju/juju/environs"
+	"github.com/juju/juju/environs/config"
+	"github.com/juju/juju/environs/configstore"
+)
+
+// CreateCommand calls the API to create a new environment.
+type CreateCommand struct {
+	envcmd.EnvCommandBase
+	api CreateEnvironmentAPI
+	// These attributes are exported only for testing purposes.
+	Name string
+	// TODO: owner string
+	ConfigFile cmd.FileVar
+	ConfValues map[string]string
+}
+
+const createEnvHelpDoc = `
+If configuration values are passed by both extra command line arguments and
+the --config option, the command line args take priority.
+
+This command will create another environment within the current Juju
+Environment Server. The provider has to match, and the environment config must
+specify all the required configuration values for the provider. In the cases
+of ‘ec2’ and ‘openstack’, the same environment variables are checked for the
+access and secret keys.
+`
+
+func (c *CreateCommand) Info() *cmd.Info {
+	return &cmd.Info{
+		Name:    "create",
+		Args:    "<name> [key=[value] ...]",
+		Purpose: "create an environment within the Juju Environment Server",
+		Doc:     strings.TrimSpace(createEnvHelpDoc),
+	}
+}
+
+func (c *CreateCommand) SetFlags(f *gnuflag.FlagSet) {
+	// TODO: support creating environments for someone else when we work
+	// out how to have the other user login and start using the environement.
+	// f.StringVar(&c.owner, "owner", "", "the owner of the new environment if not the current user")
+	f.Var(&c.ConfigFile, "config", "path to yaml-formatted file containing environment config values")
+}
+
+func (c *CreateCommand) Init(args []string) error {
+	if len(args) == 0 {
+		return errors.New("environment name is required")
+	}
+	c.Name, args = args[0], args[1:]
+
+	values, err := keyvalues.Parse(args, true)
+	if err != nil {
+		return err
+	}
+	c.ConfValues = values
+	return nil
+}
+
+type CreateEnvironmentAPI interface {
+	Close() error
+	ConfigSkeleton(provider, region string) (params.EnvironConfig, error)
+	CreateEnvironment(owner string, account, config map[string]interface{}) (params.Environment, error)
+}
+
+func (c *CreateCommand) getAPI() (CreateEnvironmentAPI, error) {
+	if c.api != nil {
+		return c.api, nil
+	}
+	root, err := c.NewAPIRoot()
+	if err != nil {
+		return nil, err
+	}
+	return environmentmanager.NewClient(root), nil
+}
+
+func (c *CreateCommand) Run(ctx *cmd.Context) (err error) {
+	client, err := c.getAPI()
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	// Create the configstore entry and write it to disk, as this will error
+	// if one with the same name already exists IF we are creating an
+	// environment for ourselves.
```

Comment:
I dont' quite understand this comment. The config is being written to disk because it will error if one with same name already exists?